### PR TITLE
C++: Support `using enum` declarations.

### PR DESCRIPTION
**Pull request by sashabu**
Previously, using entries were classified into using-declarations and using-directives based on whether they refer to a namespace or not.

With `using enum` declarations, we can no longer use the target to distinguish using-declarations and using-enum-declarations. E.g. the following:

```
namespace foo {
enum class Bar { one, two };
}
using foo::Bar;  // Brings `Bar` into scope
using enum foo::Bar;  // Brings `one` and `two` into scope.
using foo::Bar::one; using foo::Bar::two;  // Brings `one` and `two` into scope with different syntax.
```

Therefore we add a kind column. For consistency and simplicity, using-directives are also migrated to a separate kind.

### CODE REVIEW

File: cpp/downgrades/25e365d1e8147df0f759b604f96eb4bffea48271/usings.ql [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+class Location extends @location_default {
+  string toString() { none() }
+}
+
+from UsingEntry u, Element target, Location loc, int kind
+where
+  usings(u, target, loc, kind) and
+  kind != 3
+select u, target, loc
```
Comment (jketema) [inline cpp/downgrades/25e365d1e8147df0f759b604f96eb4bffea48271/usings.ql:17]: Nothing in the database can refer to a `UsingEntry`, right?

File: cpp/ql/lib/upgrades/3d35dd6b50edfc540c14c6757e0c7b3c5b7b04dd/usings.ql [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+class Location extends @location_default {
+  string toString() { none() }
+}
+
+from UsingEntry u, Element target, Location loc, int kind
+where
+  usings(u, target, loc) and
+  if target instanceof @namespace then kind = 2 else kind = 1
+select u, target, loc, kind
```
Comment (jketema) [inline cpp/ql/lib/upgrades/3d35dd6b50edfc540c14c6757e0c7b3c5b7b04dd/usings.ql:17]: I assume you did some basic testing to ensure this works as expected?

**Review by jketema — COMMENTED** [review]
Two small questions, otherwise this looks fine to me.

File: cpp/downgrades/25e365d1e8147df0f759b604f96eb4bffea48271/usings.ql (hunk 1 above)
Comment (sashabu) [reply cpp/downgrades/25e365d1e8147df0f759b604f96eb4bffea48271/usings.ql:17]: Fixed in https://github.com/github/codeql/pull/17006/commits/af562f15f3099a4291c281d85c1881fb401438d2. (Don't think we need to do anything on the upgrade side.)

File: cpp/ql/lib/upgrades/3d35dd6b50edfc540c14c6757e0c7b3c5b7b04dd/usings.ql (hunk 2 above)
Comment (sashabu) [reply cpp/ql/lib/upgrades/3d35dd6b50edfc540c14c6757e0c7b3c5b7b04dd/usings.ql:17]: Yes, I spot checked the version you reviewed by creating a database for the usings test with the before/after CLI and running the queries with after/before CLIs respectively. I'm doing the same for the `using_container` fix now.

Comment (sashabu) [reply cpp/ql/lib/upgrades/3d35dd6b50edfc540c14c6757e0c7b3c5b7b04dd/usings.ql:17]: Just confirmed all looks good with the additional typo fix in https://github.com/github/codeql/pull/17006/commits/3defc8b5de03fb9219b0cacf520d69c1a597604e.

**Pull request merged**